### PR TITLE
Added Neopixel Example

### PR DESCRIPTION
**Pull request by curuvar**
## Summary

Example app using the neopixel driver that displays a rainbow pattern.

## Impact

Should not impact other apps.

## Testing

Tested app with single and multiple pixel chains with various update speeds.

### CODE REVIEW

File: examples/neopixel/neopixel_main.c [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * hello_main
+ ****************************************************************************/
+
+int neopixel_main(int argc, FAR char *argv[])
```
Comment (xiaoxiang781216) [inline examples/neopixel/neopixel_main.c:239]: change main, the build system will rename it for you automatically.

File: examples/neopixel/neopixel_main.c [hunk 2]
```diff
@@ -38,6 +38,196 @@
 
 #include <nuttx/leds/neopixel.h>
 
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct neo_config_s
+{
+  char      * path;
```
Comment (xiaoxiang781216) [inline examples/neopixel/neopixel_main.c:47]: remove space before path, and add FAR

File: examples/neopixel/neopixel_main.c [hunk 3]
```diff
@@ -38,6 +38,196 @@
...
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: help
+ ****************************************************************************/
+
+static void help(FAR struct neo_config_s  * conf)
```
Comment (xiaoxiang781216) [inline examples/neopixel/neopixel_main.c:73]: remove space before conf

File: examples/neopixel/neopixel_main.c [hunk 4]
```diff
@@ -38,6 +38,196 @@
...
+         "Default: %d us Current: %d us\n",
+         20000, conf->delay);
+}
+
+/****************************************************************************
+ * Name: set_devpath
+ ****************************************************************************/
+
+static void set_devpath(FAR struct neo_config_s  * conf,
+                        FAR const char           * devpath)
```
Comment (xiaoxiang781216) [inline examples/neopixel/neopixel_main.c:104]: ditto

File: examples/neopixel/neopixel_main.c [hunk 5]
```diff
@@ -38,6 +38,196 @@
...
+  ret = arg_string(arg, &string);
+  *value = strtol(string, NULL, 10);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: parse_args
+ ****************************************************************************/
+
+static void parse_args(FAR struct neo_config_s  * conf,
```
Comment (xiaoxiang781216) [inline examples/neopixel/neopixel_main.c:156]: ditto

File: examples/neopixel/neopixel_main.c [hunk 6]
```diff
@@ -38,6 +38,196 @@
...
+                       int argc,
+                       FAR char **argv)
+{
+  FAR char  * ptr;
+  FAR char  * str;
```
Comment (xiaoxiang781216) [inline examples/neopixel/neopixel_main.c:161]: ditto

File: examples/neopixel/neopixel_main.c [hunk 7]
```diff
@@ -48,48 +238,53 @@
...
 int neopixel_main(int argc, FAR char *argv[])
 {
-  printf("Hello, World, from neopixel!!\n");
-
-  char      * path      = CONFIG_EXAMPLE_NEOPIXEL_DEFAULT_DEV;
-  int         loops     =   1024; /* all colors 4 times */
-  int         leds      =      1;
-  int         delay     =  20000; /* µs -- ~50Hz */
   uint32_t  * buffer;
   uint32_t  * bp;
```
Comment (xiaoxiang781216) [inline examples/neopixel/neopixel_main.c:242]: add FAR and remove space before variables

File: examples/neopixel/neopixel_main.c (hunk 1 above)
Comment (curuvar) [reply examples/neopixel/neopixel_main.c:239]: I am not sure what you mean by "change main".

Comment (xiaoxiang781216) [reply examples/neopixel/neopixel_main.c:239]: neopixel_main->main

File: examples/neopixel/Make.defs [hunk 8]
```diff
@@ -0,0 +1,23 @@
+############################################################################
+# apps/examples/hello/Make.defs
```
Comment (gustavonihei) [inline examples/neopixel/Make.defs:2]:
```suggestion
# apps/examples/neopixel/Make.defs
```

File: examples/neopixel/neopixel_main.c [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: help
+ ****************************************************************************/
+
+static void help(FAR struct neo_config_s  *conf)
```
Comment (gustavonihei) [inline examples/neopixel/neopixel_main.c:73]:
```suggestion
static void help(FAR struct neo_config_s *conf)
```
nit: double whitespace

File: examples/neopixel/neopixel_main.c [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+          exit(0);
+        }
+
+      switch (ptr[1])
+        {
+          case 'l':
+            nargs = arg_decimal(&argv[index], &value);
+            if (value < 1 || value > INT_MAX)
+              {
+                printf("Led count out of range: %ld\n", value);
```
Comment (gustavonihei) [inline examples/neopixel/neopixel_main.c:181]:
```suggestion
                fprintf(stderr, "Led count out of range: %ld\n", value);
```

File: examples/neopixel/neopixel_main.c [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+
+            conf->leds = (uint32_t)value;
+            index += nargs;
+            break;
+
+          case 'r':
+            nargs = arg_decimal(&argv[index], &value);
+            if (value < 1 || value > INT_MAX)
+              {
+                printf("Repeat out of range: %ld\n", value);
```
Comment (gustavonihei) [inline examples/neopixel/neopixel_main.c:193]:
```suggestion
                fprintf(stderr, "Repeat out of range: %ld\n", value);
```

File: examples/neopixel/neopixel_main.c [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+            conf->delay = (int)value;
+            index += nargs;
+            break;
+
+          case 'h':
+            help(conf);
+            exit(0);
+
+          default:
+            printf("Unsupported option: %s\n", ptr);
```
Comment (gustavonihei) [inline examples/neopixel/neopixel_main.c:224]:
```suggestion
            fprintf(stderr, "Unsupported option: %s\n", ptr);
```

File: .gitignore [hunk 13]
```diff
@@ -31,3 +31,5 @@ Make.dep
 .context
 # nuttx/$(TOPDIR)/Makefile.[unix|win]::$(DIRLINKS_EXTERNAL_DIRS)
 .dirlinks
+.vscode
+.DS_Store
```
[the remark this answers is not in the document]
Comment (curuvar) [reply .gitignore:35]: The .vscode is added by VSCode and surely should not be saved in the repository.  Likewise .DS_Store is added to folder by MacOS and is also unwanted.

**Comment by xiaoxiang781216** [comment]
@curuvar could you squash all patch into one by:
git rebase --interactive HEAD~5

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
why not squash the rename patch into the first?

File: .gitignore (hunk 13 above)
Comment (pkarashchenko) [inline .gitignore:35]:
```suggestion
.DS_Store

```

**Review by xiaoxiang781216 — APPROVED** [review]

File: examples/ws2812/Makefile [hunk 14]
```diff
@@ -0,0 +1,32 @@
+############################################################################
+# apps/apps/examples/ws2812/Makefile
```
Comment (gustavonihei) [inline examples/ws2812/Makefile:2]:
```suggestion
# apps/examples/ws2812/Makefile
```

**Comment by xiaoxiang781216** [comment]
@curuvar it's better to squash the change for comment in the main patch .

**Comment by curuvar** [comment]
@xiaoxiang781216, yes, I have been trying to remember to squash before I push changes.  It turns out the change in question was make when I clicked the "Commit Suggestion" button on gustavonihei's comment.  Is it best to avoid this practice, or is there some way to have it squash from the GitHub site itself?

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216, yes, I have been trying to remember to squash before I push changes. It turns out the change in question was make when I clicked the "Commit Suggestion" button on gustavonihei's comment. Is it best to avoid this practice, or is there some way to have it squash from the GitHub site itself?

I normally update the change locally instead from the web interface. Another way is enable Squash and Merge:

image

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**